src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,394 @@
+﻿using Autodesk.DesignScript.Interfaces;
+using Dynamo.Logging;
+using Dynamo.PythonServices;
+using Python.Runtime;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+namespace DSCPython
+{
+    internal class DSCPythonCodeCompletionProviderCore : PythonCodeCompletionProviderCommon, ILogSource, IDisposable
+    {
+        #region Private members
+
+        /// <summary>
+        /// The engine used for autocompletion.  This essentially keeps
+        /// track of the state of the editor, allowing access to variable types and
+        /// imported symbols.
+        /// </summary>
+        private PyScope scope;
+
+        private string uniquePythonScopeName = "uniqueScope";
+
+        /// <summary>
+        /// The scope used by the engine.  This is where all the loaded symbols
+        /// are stored.  It's essentially an environment dictionary.
+        /// </summary>
+        private PyScope Scope
+        {
+            get { return scope; }
+            set { scope = (PyScope)value; }
+        }
+
+        private PyScope CreateUniquePythonScope()
+        {
+            PyScopeManager.Global.TryGet(uniquePythonScopeName, out PyScope Scope);
+
+            if (Scope == null)
+            {
+                Scope = Py.CreateScope(uniquePythonScopeName);
+                Scope.Exec(@"
+import clr
+import sys
+clr.setPreload(True)
+");
+            }
+
+            return Scope;
+        }
+
+        private object ExecutePythonScriptCode(string code)
+        {
+            Python.Included.Installer.SetupPython().Wait();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            using (Py.GIL())
+            {
+                var result = Scope.Eval(code);
+                return CPythonEvaluator.OutputMarshaler.Marshal(result);
+            }
+        }
+
+        /// <summary>
+        /// List all of the members in a PythonObject
+        /// </summary>
+        /// <param name="pyObject">A reference to the module</param>
+        /// <param name="name">The name of the module</param>
+        /// <returns>A list of completion data for the module</returns>
+        private IEnumerable<Tuple<string, string, bool, ExternalCodeCompletionType>> EnumerateMembersForPyObj(object pyObject, string name)
+        {
+            var items = new List<Tuple<string, string, bool, ExternalCodeCompletionType>>();
+            var d = pyObject as PyObject;
+            if (d == null)
+            {
+                return items;
+            }
+
+            foreach (var member in d.Dir())
```

Comment:
I wonder what the difference is between dir and dynamicMembers

Reply:
```
        /// <summary>
        /// Returns the enumeration of all dynamic member names.
        /// </summary>
        /// <remarks>
        /// This method exists for debugging purposes only.
        /// </remarks>
        /// <returns>A sequence that contains dynamic member names.</returns>
        public override IEnumerable<string> GetDynamicMemberNames()
        {
            foreach (PyObject pyObj in Dir())
            {
                yield return pyObj.ToString();
            }
        }
        ```